library/index.c

```diff
@@ -380,30 +395,63 @@ static struct eblob_disk_control *eblob_find_on_disk(struct eblob_backend *b,
 			eblob_dump_id(search_end->key.id), num);
 
 	if (!sorted_orig)
-		goto out;
+		goto err_out_free_index;
 
 	st->bsearch_found++;
 
+	/*
+	 * Sorted index may contain range of keys with the same key.
+	 * Some keys in that range may be marked as removed.
+	 * Do forward linear search (reading block by block) in range of keys,
+	 * until reached range's end or found existing key.
+	 */
 	sorted = sorted_orig;
-	while (sorted < end && eblob_disk_control_sort(sorted, dc) == 0) {
+	end = search_end;
+	saved_hdr_block_offset = hdr_block_offset;
+	while (eblob_disk_control_sort(sorted, dc) == 0) {
 		if (callback(sorted, dc)) {
 			found = sorted;
 			break;
 		}
 		st->additional_reads++;
-		sorted++;
+
+		if (++sorted > end) {
+			/* read next block of headers */
+			hdr_block_offset += hdr_block_size;
+			if (hdr_block_offset >= bctl->sort.size)
+				break;
+
+			num = (bctl->sort.size - hdr_block_offset) / hdr_size;
+
+			if (num > b->cfg.index_block_size)
+				num = b->cfg.index_block_size;
+
+			hdr_block_size = num * hdr_size;
+			sorted = hdr_block;
+			end = sorted + (num - 1);
+
+			read_err = __eblob_read_ll(bctl->sort.fd, hdr_block, hdr_block_size, hdr_block_offset);
```

Comment:
Please add error log after **every** invocation of this function and put as much information as possible: index number, total size, idnex size, position, read position, size and so on. As much information as possible.